vulnerabilities/package_managers.py

```diff
@@ -489,3 +500,93 @@ async def fetch(self, pkg, session):
             pass
 
         self.cache[pkg] = versions
+
+
+class GoproxyVersionAPI(VersionAPI):
+
+    package_type = "golang"
+    module_name_by_package_name = {}
+
+    @staticmethod
+    def trim_url_path(url_path: str) -> Optional[str]:
+        """
+        github advisories for golang is using package names(e.g.
+        https://github.com/advisories/GHSA-jp4j-47f9-2vc3),
+        yet goproxy works with module names(see
+        https://golang.org/ref/mod#goproxy-protocol).
+        this method removes the last part of a package path, and returns the
+        remaining as the module name. For example: trim_url_path(
+        "https://github.com/xx/a/b") returns "https://github.com/xx/a"
+        """
+        # some advisories contains this prefix in package name, e.g. https://github.com/advisories/GHSA-7h6j-2268-fhcm
+        if url_path.startswith("https://pkg.go.dev/"):
+            url_path = url_path.removeprefix("https://pkg.go.dev/")
+        parts = url_path.split("/")
+        if len(parts) >= 2:
+            return "/".join(parts[:-1])
+        else:
+            return None
+
+    @staticmethod
+    def escape_path(path: str) -> str:
+        """
+        escape uppercase in module/version name.
+        """
+        escaped_path = ""
+        for c in path:
+            if c >= "A" and c <= "Z":
+                # replace uppercase with !lowercase
+                escaped_path += "!" + chr(ord(c) + ord("a") - ord("A"))
+            else:
+                escaped_path += c
+        return escaped_path
+
+    @staticmethod
+    async def parse_version_info(
+        version_info: str, escaped_pkg: str, session: ClientSession
+    ) -> Optional[Version]:
+        v = version_info.split()
+        if len(v) > 0:
+            value = v[0]
+            if len(v) > 1:
+                release_date = parse_datetime(v[1])
+            else:
+                escaped_ver = GoproxyVersionAPI.escape_path(value)
+                try:
+                    response = await session.request(
+                        method="GET",
+                        url=f"https://proxy.golang.org/{escaped_pkg}/@v/{escaped_ver}.info",
+                    )
+                    resp_json = await response.json()
+                    release_date = parse_datetime(resp_json.get("Time", ""))
+                except:
+                    release_date = None
```

Comment:
what if there was a 500 or a 404? ... catching all exceptions feels a bit brutal to me.

Reply:
I thought release date was optional, so I just set it to None when having problems querying it.

Reply:
@sify21 wrt. to errors, we need to have a centralized way to make network requests and deal with errors and retries so what you can do for now is create an issue as a reminder to revisit this later to properly handle errors.
But in anycase, do not silently pass all exceptions there. Instead do some logging at the minimum for now.

Reply:
I added some code to log error messages, and created the issue here https://github.com/nexB/vulnerablecode/issues/619